sample_consensus/include/pcl/sample_consensus/impl/lmeds.hpp

```diff
@@ -96,22 +97,36 @@ pcl::LeastMedianSquares<PointT>::computeModel (int debug_verbosity_level)
       ++skipped_count;
       continue;
     }
+    // Move all NaNs in distances to the end
+    const auto new_end = (sac_model_->getInputCloud()->is_dense ? distances.end() : std::partition (distances.begin(), distances.end(), [](double d){return !std::isnan (d);}));
+    const auto nr_valid_dists = std::distance (distances.begin (), new_end);
 
-    std::sort (distances.begin (), distances.end ());
     // d_cur_penalty = median (distances)
-    std::size_t mid = sac_model_->getIndices ()->size () / 2;
-    if (mid >= distances.size ())
+    const std::size_t mid = nr_valid_dists / 2;
+    PCL_DEBUG ("[pcl::LeastMedianSquares::computeModel] There are %lu valid distances remaining after removing NaN values.\n", nr_valid_dists);
+    if (nr_valid_dists == 0)
     {
       //iterations_++;
       ++skipped_count;
       continue;
     }
 
     // Do we have a "middle" point or should we "estimate" one ?
-    if (sac_model_->getIndices ()->size () % 2 == 0)
-      d_cur_penalty = (sqrt (distances[mid-1]) + sqrt (distances[mid])) / 2;
+    if ((nr_valid_dists % 2) == 0)
+    {
+      // Looking at two values instead of one probably doesn't matter because they are mostly barely different, but let's do it for accuracy's sake
+      std::nth_element (distances.begin (), distances.begin () + (mid - 1), new_end);
```

Comment:
Just a side: Would it make sense to move `nth_element` till `mid` out of the 2 branches? Then `tmp2` can be min_element from `begin` to `mid - 1`. This method separates the 2 cases so this is fine too.

Reply:
I think it is good like it is now, because the separation might help understanding it.